File: apps/web/src/api/project-summary.ts
// 프로젝트 요약·활동 API 클라이언트 — Zod 스키마 + fetch + 델타 계산 순수 함수 (Jira 패리티 J4)
import { z } from 'zod'
import { apiFetch, ApiError } from './client'
import { changeItemSchema } from './changelog'

// ─────────────────────────────────────────────────────────────────────────────
// Zod 스키마 — 백엔드 DTO(ProjectSummaryResponse/ProjectActivityResponse,
// backend/modules/issue-tracking/.../summary/web/dto/*.kt)와 필드명·타입 1:1 grep 대조.
// @JsonInclude(NON_NULL) 이 걸린 필드는 키 자체가 빠져 오므로 .nullish() 를 쓴다.
// ─────────────────────────────────────────────────────────────────────────────

/**
 * 현재 창과 직전 창의 값 쌍 Zod 스키마.
 * 백엔드 WindowCountResponse DTO와 1:1 대응.
 * - current: 최근 7일 값
 * - previous: 직전 7일(8~14일 전) 값 — 카드의 델타 문구 재료
 */
export const windowCountSchema = z.object({
  current: z.number(),
  previous: z.number(),
})

/** 창 카운트 쌍 타입 */
export type WindowCount = z.infer<typeof windowCountSchema>

/**
 * 최근 7일 카드 3종 Zod 스키마.
 * 백엔드 RecentCountsResponse DTO와 1:1 대응.
 * - completed: 상태 이력의 DONE 진입 시각 기준(`updated_at` 기준이 아니다)
 */
export const recentCountsSchema = z.object({
  windowDays: z.number(),
  completed: windowCountSchema,
  updated: windowCountSchema,
  created: windowCountSchema,
})

/** 최근 7일 카드 3종 타입 */
export type RecentCounts = z.infer<typeof recentCountsSchema>

/**
 * 마감 예정·지연 카드 Zod 스키마.
 * 백엔드 UpcomingCountsResponse DTO와 1:1 대응. 둘 다 미완료 이슈만 센다.
 */
export const upcomingCountsSchema = z.object({
  windowDays: z.number(),
  due: z.number(),
  overdue: z.number(),
})

/** 마감 예정·지연 카드 타입 */
export type UpcomingCounts = z.infer<typeof upcomingCountsSchema>

/** 상태 카테고리 — 백엔드 StatusCategory enum 3값과 1:1 */
export const statusCategorySchema = z.enum(['TODO', 'IN_PROGRESS', 'DONE'])

/** 상태 카테고리 타입 */
export type StatusCategory = z.infer<typeof statusCategorySchema>

/**
 * 상태 개요 한 조각 Zod 스키마.
 * 백엔드 StatusSliceResponse DTO와 1:1 대응.
 * - statusName: 워크플로우 스킴 해석 실패 시 키가 빠진다 → 화면은 statusKey 로 폴백
 */
export const statusSliceSchema = z.object({
  statusKey: z.string(),
  statusName: z.string().nullish(),
  category: statusCategorySchema,
  count: z.number(),
})

/** 상태 개요 조각 타입 */
export type StatusSlice = z.infer<typeof statusSliceSchema>

/**
 * 우선순위 분포 한 조각 Zod 스키마.
 * 백엔드 PrioritySliceResponse DTO와 1:1 대응.
 * - priorityName: SMALLINT 제약 밖 값이면 키가 빠진다 → 화면은 숫자로 폴백
 */
export const prioritySliceSchema = z.object({
  priority: z.number(),
  priorityName: z.string().nullish(),
  count: z.number(),
})

/** 우선순위 분포 조각 타입 */
export type PrioritySlice = z.infer<typeof prioritySliceSchema>

/**
 * 작업 유형 분포 한 조각 Zod 스키마.
 * 백엔드 TypeSliceResponse DTO와 1:1 대응. 두 필드 모두 non-null 이다.
 */
export const typeSliceSchema = z.object({
  typeKey: z.string(),
  typeName: z.string(),
  count: z.number(),
})

/** 작업 유형 분포 조각 타입 */
export type TypeSlice = z.infer<typeof typeSliceSchema>

/**
 * 담당자 분포 한 조각 Zod 스키마.
 * 백엔드 AssigneeSliceResponse DTO와 1:1 대응.
 * - assigneeId 부재 = 미할당 묶음. assigneeName 부재 = 표시명 조회 실패
 */
export const assigneeSliceSchema = z.object({
  assigneeId: z.string().nullish(),
  assigneeName: z.string().nullish(),
  count: z.number(),
})

/** 담당자 분포 조각 타입 */
export type AssigneeSlice = z.infer<typeof assigneeSliceSchema>

/**
 * 프로젝트 요약 응답 Zod 스키마.
 * GET /api/v1/projects/{projectKey}/summary 의 data 필드 형식.
 */
export const projectSummarySchema = z.object({
  projectKey: z.string(),
  recent: recentCountsSchema,
  upcoming: upcomingCountsSchema,
  statusOverview: z.array(statusSliceSchema),
  priorityBreakdown: z.array(prioritySliceSchema),
  typesOfWork: z.array(typeSliceSchema),
  teamWorkload: z.array(assigneeSliceSchema),
})

/** 프로젝트 요약 응답 타입 */
export type ProjectSummary = z.infer<typeof projectSummarySchema>

/**
 * 활동 피드 항목 하나 Zod 스키마.
 * 백엔드 ProjectActivityEntryResponse DTO와 1:1 대응.
 *
 * `items` 는 이슈 단건 changelog 와 같은 [changeItemSchema] 를 **재사용**한다 — 모양이 같은
 * 스키마를 하나 더 두면 백엔드가 필드를 늘릴 때 한쪽만 따라간다.
 */
export const projectActivityEntrySchema = z.object({
  issueKey: z.string(),
  actorId: z.string().nullish(),
  actorName: z.string().nullish(),
  createdAt: z.string(),
  items: z.array(changeItemSchema),
})

/** 활동 피드 항목 타입 */
export type ProjectActivityEntry = z.infer<typeof projectActivityEntrySchema>

/**
 * 프로젝트 활동 피드 응답 Zod 스키마.
 * GET /api/v1/projects/{projectKey}/activity 의 data 필드 형식.
 */
export const projectActivitySchema = z.object({
  entries: z.array(projectActivityEntrySchema),
})

/** 프로젝트 활동 피드 응답 타입 */
export type ProjectActivity = z.infer<typeof projectActivitySchema>

// ─────────────────────────────────────────────────────────────────────────────
// 내부 헬퍼
// ─────────────────────────────────────────────────────────────────────────────

/** backend 응답 래퍼 `{ data: T }` 파싱 헬퍼 */
const dataResponseSchema = <T>(innerSchema: z.ZodSchema<T>) => z.object({ data: innerSchema })

// ─────────────────────────────────────────────────────────────────────────────
// API 함수
// ─────────────────────────────────────────────────────────────────────────────

/**
 * 프로젝트 요약 집계를 조회한다.
 *
 * GET /api/v1/projects/{projectKey}/summary → 200 { data: ProjectSummary }
 * 창은 백엔드 고정(최근 7일 · 상태 개요 DONE 만 최근 2주)이라 쿼리 파라미터가 없다.
 *
 * @param projectKey 조회할 프로젝트 키 (예: "BTS")
 * @returns 프로젝트 요약 응답
 * @throws ApiError(401) 미인증
 * @throws ApiError(403) 프로젝트 BROWSE 권한 없음
 */
export async function fetchProjectSummary(projectKey: string): Promise<ProjectSummary> {
  const res = await apiFetch(`/api/v1/projects/${projectKey}/summary`, { method: 'GET' })
  if (!res.ok) {
    const errorBody: unknown = await res.json().catch(() => ({}))
    throw new ApiError(res.status, errorBody)
  }
  const raw: unknown = await res.json()
  return dataResponseSchema(projectSummarySchema).parse(raw).data
}

/**
 * 프로젝트 활동 피드를 최신순으로 조회한다.
 *
 * GET /api/v1/projects/{projectKey}/activity?limit={limit} → 200 { data: ProjectActivity }
 * 백엔드가 limit 을 1~50 으로 강제하며 범위 밖은 400 이다. 커서는 없다 — 위젯이 페이지네이션하지 않는다.
 *
 * @param projectKey 조회할 프로젝트 키
 * @param limit 최대 항목 수. 기본 20
 * @returns 프로젝트 활동 피드 응답
 * @throws ApiError(400) limit 이 1~50 밖
 * @throws ApiError(401) 미인증
 * @throws ApiError(403) 프로젝트 BROWSE 권한 없음
 */
export async function fetchProjectActivity(
  projectKey: string,
  limit = 20,
): Promise<ProjectActivity> {
  const res = await apiFetch(`/api/v1/projects/${projectKey}/activity?limit=${limit}`, {
    method: 'GET',
  })
  if (!res.ok) {
    const errorBody: unknown = await res.json().catch(() => ({}))
    throw new ApiError(res.status, errorBody)
  }
  const raw: unknown = await res.json()
  return dataResponseSchema(projectActivitySchema).parse(raw).data
}

// ─────────────────────────────────────────────────────────────────────────────
// 순수 함수
// ─────────────────────────────────────────────────────────────────────────────

/** 카드 델타 판정 결과 — 화면이 부호·색을 고르는 재료 */
export interface WindowDelta {
  /** current - previous */
  readonly diff: number
  /** 증감 방향. 0 이면 'flat' */
  readonly direction: 'up' | 'down' | 'flat'
}

/**
 * 창 카운트 쌍에서 직전 창 대비 델타를 계산하는 순수 함수.
 *
 * 화면 세 곳(카드 3종)이 각자 빼기를 하면 부호 규칙이 갈린다. 한 자리에 모은다.
 *
 * @param count 현재·직전 창 값 쌍
 * @returns 차이와 방향
 */
export function resolveWindowDelta(count: WindowCount): WindowDelta {
  const diff = count.current - count.previous
  if (diff > 0) return { diff, direction: 'up' }
  if (diff < 0) return { diff, direction: 'down' }
  return { diff: 0, direction: 'flat' }
}

/**
 * 요약 응답에 그릴 분포가 하나도 없는지 판별하는 순수 함수.
 *
 * 카드(최근 7일·마감)는 0 이어도 「0」을 보여주는 것이 정보다. 분포 4종이 전부 비었을 때만
 * 「아직 이슈가 없다」로 본다.
 *
 * @param summary 프로젝트 요약 응답
 * @returns 분포 4종이 모두 비어 있으면 true
 */
export function isProjectSummaryEmpty(summary: ProjectSummary): boolean {
  return (
    summary.statusOverview.length === 0 &&
    summary.priorityBreakdown.length === 0 &&
    summary.typesOfWork.length === 0 &&
    summary.teamWorkload.length === 0
  )
}

File: backend/modules/issue-tracking/src/main/kotlin/com/bts/issue/summary/domain/ProjectSummary.kt
// 프로젝트 요약 화면 집계 결과 도메인 모델 — Jira 클라우드 Summary view 패리티

package com.bts.issue.summary.domain

import com.bts.issue.statushistory.StatusCategory
import java.util.UUID

/**
 * 프로젝트 요약 화면 한 벌의 집계 결과.
 *
 * Jira 클라우드 Summary view 와 같은 스펙을 목표로 한다.
 *
 * @property projectKey 집계 대상 프로젝트 키.
 * @property recent 상단 카드 3종(완료·업데이트·생성) — 최근 7일과 직전 7일.
 * @property upcoming 상단 카드 「마감 예정」 — 향후 7일 + 지연.
 * @property statusOverview 상태 개요. **DONE 카테고리만 최근 2주 특례가 걸린다**([StatusSlice]).
 * @property priorityBreakdown 우선순위 분포 — 활성·가시 이슈 전량.
 * @property typesOfWork 작업 유형 분포 — 활성·가시 이슈 전량.
 * @property teamWorkload 담당자 분포 — 활성·가시 이슈 전량.
 */
data class ProjectSummary(
    val projectKey: String,
    val recent: RecentCounts,
    val upcoming: UpcomingCounts,
    val statusOverview: List<StatusSlice>,
    val priorityBreakdown: List<PrioritySlice>,
    val typesOfWork: List<TypeSlice>,
    val teamWorkload: List<AssigneeSlice>,
)

/**
 * 현재 창과 직전 창의 값 쌍. 화면이 델타(「+4」·「변동 없음」)를 그리는 재료다.
 *
 * @property current 최근 7일 값.
 * @property previous 직전 7일(8~14일 전) 값.
 */
data class WindowCount(
    val current: Long,
    val previous: Long,
)

/**
 * 최근 7일 활동 카드 3종.
 *
 * @property windowDays 창 길이(일). 항상 7.
 * @property completed 완료 — **상태 이력에서 DONE 카테고리로 진입한 시각** 기준(D1).
 * @property updated 업데이트 — `issues.updated_at` 기준.
 * @property created 생성 — `issues.created_at` 기준.
 */
data class RecentCounts(
    val windowDays: Int,
    val completed: WindowCount,
    val updated: WindowCount,
    val created: WindowCount,
)

/**
 * 마감 관련 카드.
 *
 * 둘 다 **미완료 이슈만** 센다 — 이미 끝난 일은 마감을 앞두고 있지도, 지연되지도 않았다.
 *
 * @property windowDays 창 길이(일). 항상 7.
 * @property due 오늘 포함 향후 7일 안에 마감인 미완료 이슈 수.
 * @property overdue 마감일이 오늘보다 이전인 미완료 이슈 수.
 */
data class UpcomingCounts(
    val windowDays: Int,
    val due: Long,
    val overdue: Long,
)

/**
 * 상태 개요 한 조각.
 *
 * ### DONE 2주 특례
 * [category] 가 [StatusCategory.DONE] 인 조각은 **최근 2주 안에 완료된 이슈만** 센다.
 * Jira 클라우드 원문 — "Only items that have been completed in the last two weeks will appear
 * in Done". 다른 카테고리와 [PrioritySlice]·[TypeSlice]·[AssigneeSlice] 에는 적용되지 않는다.
 *
 * @property statusKey 워크플로우 상태 키.
 * @property statusName 표시명. 워크플로우 스킴을 해석하지 못하면 null — 화면이 키로 폴백한다.
 * @property category 상태 카테고리.
 * @property count 이슈 수.
 */
data class StatusSlice(
    val statusKey: String,
    val statusName: String?,
    val category: StatusCategory,
    val count: Long,
)

/**
 * 우선순위 분포 한 조각.
 *
 * @property priority 우선순위 값(1~5). 표시명은 DTO 매퍼가 `IssuePriority` 로 해석해 싣는다.
 * @property count 이슈 수.
 */
data class PrioritySlice(
    val priority: Int,
    val count: Long,
)

/**
 * 작업 유형 분포 한 조각.
 *
 * @property typeKey 이슈 유형 키(예: `story`).
 * @property typeName 이슈 유형 표시명.
 * @property count 이슈 수.
 */
data class TypeSlice(
    val typeKey: String,
    val typeName: String,
    val count: Long,
)

/**
 * 담당자 분포 한 조각.
 *
 * @property assigneeId 담당자 UUID. null 이면 미할당 버킷.
 * @property assigneeName 표시명. 미할당이거나 identity-access 조회가 실패하면 null(graceful degrade).
 * @property count 이슈 수.
 */
data class AssigneeSlice(
    val assigneeId: UUID?,
    val assigneeName: String?,
    val count: Long,
)
